### PR TITLE
expression: handle ENUM in builtin function Values()

### LINKED ISSUE #9165: enum + insert on duplicate update > data truncate
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
CREATE TABLE `table1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `f2` enum('cluster','default','common') DEFAULT 'common',
  `f3` varchar(1000) DEFAULT NULL,
  `f4` varchar(255) DEFAULT NULL COMMENT 'group by f3',
  `f5` int(11) DEFAULT NULL ,
  `f6` int(11) DEFAULT NULL ,
  `f7` int(11) DEFAULT '100' ,
  `f8` int(11) DEFAULT NULL,
  `f9` int(11) DEFAULT NULL,
  `f10` varchar(50) DEFAULT NULL,
  `f11` varchar(200) DEFAULT NULL,
  `f12` varchar(500) DEFAULT 'email,hotchat,message' ,
  `f13` int(11) DEFAULT '1',
  `f14` datetime DEFAULT NULL,
  `f15` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin;

sql xxx:
insert into `test`.`table1`(`f2` , `f3` , `f4` , `f5` , `f6` , `f7` , `f8` , `f9` , `f10` , `f11` , `f12` , `f13` , `f14` , `f15` , `id`) values (1 , '{"group":"hubble","metric":"cpu.load.5min.per.core,agent.alive,proc.used.monitor"}' , '' , 10 , 2 , 1 , 3 , 1 , 'weiwei' , 'weiwei' , 'email,message,hotchat' , 1 , NULL , '2019-01-23 15:36:25' , 34) on duplicate key update `f2`=values(`f2`) , `f3`=values(`f3`) , `f4`=values(`f4`) , `f5`=values(`f5`) , `f6`=values(`f6`) , `f7`=values(`f7`) , `f8`=values(`f8`) , `f9`=values(`f9`) , `f10`=values(`f10`) , `f11`=values(`f11`) , `f12`=values(`f12`) , `f13`=values(`f13`) , `f14`=values(`f14`) , `f15`=values(`f15`) , `id`=values(`id`);

its OK

replace sql xxx the value of type to 2 ,execute sql xxx again.

its wrong

2. What did you expect to see?
select * from table1;
the value of type is default

3. What did you see instead?
he value of type is cluster

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
v2.1.0, just specific tidb cluster can recur.
if remove the field f2, the error dissapear.
show warnings :  Data Truncated

### PR DESCRIPTION
**Pull request by zz-jason**
Signed-off-by: Jian Zhang <zjsariel@gmail.com>

### What problem does this PR solve? 

fix #9165

### What is changed and how it works?

handle ENUM in `builtinValuesStringSig`

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: expression/builtin_other.go [hunk 1]
```diff
@@ -615,13 +615,27 @@ func (b *builtinValuesStringSig) evalString(_ chunk.Row) (string, bool, error) {
 	if row.IsEmpty() {
 		return "", true, errors.New("Session current insert values is nil")
 	}
-	if b.offset < row.Len() {
-		if row.IsNull(b.offset) {
+	if b.offset >= row.Len() {
+		return "", true, errors.Errorf("Session current insert values len %d and column's offset %v don't match", row.Len(), b.offset)
+	}
+
+	if retType := b.getRetTp(); retType.Hybrid() {
```
Comment (eurekaka) [inline expression/builtin_other.go]: Add some comments for this branch?

File: expression/builtin_other.go [hunk 2]
```diff
@@ -615,13 +615,27 @@ func (b *builtinValuesStringSig) evalString(_ chunk.Row) (string, bool, error) {
...
+		val := row.GetDatum(b.offset, b.getRetTp())
```
Comment (eurekaka) [inline expression/builtin_other.go]:
```suggestion
		val := row.GetDatum(b.offset, retType)
```

File: expression/builtin_other.go [hunk 3]
```diff
@@ -615,13 +615,27 @@ func (b *builtinValuesStringSig) evalString(_ chunk.Row) (string, bool, error) {
...
-		return row.GetString(b.offset), false, nil
+		res, err := val.ToString()
+		resLen := len([]rune(res))
+		if b.ctx.GetSessionVars().StmtCtx.PadCharToFullLength && retType.Tp == mysql.TypeString && resLen < retType.Flen {
+			res = res + strings.Repeat(" ", retType.Flen-resLen)
+		}
+		return res, err != nil, err
+	}
+
+	if row.IsNull(b.offset) {
```
Comment (eurekaka) [inline expression/builtin_other.go]: Put this check before line 622 so we can remove null check in line 624?

File: expression/builtin_other.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_other.go]: done

File: expression/builtin_other.go [hunk 4]
```diff
@@ -615,13 +615,30 @@ func (b *builtinValuesStringSig) evalString(_ chunk.Row) (string, bool, error) {
...
+	if retType := b.getRetTp(); retType.Hybrid() {
+		val := row.GetDatum(b.offset, retType)
+		res, err := val.ToString()
+		if err != nil {
+			return "", true, err
 		}
-		return row.GetString(b.offset), false, nil
+
+		resLen := len([]rune(res))
+		if b.ctx.GetSessionVars().StmtCtx.PadCharToFullLength && retType.Tp == mysql.TypeString && resLen < retType.Flen {
```
Comment (lamxTyler) [inline expression/builtin_other.go]: Since the `retType` is hybird, will it be `TypeString`?

Comment (zz-jason) [reply expression/builtin_other.go]: no. updated.

**Pull request merged**